subprojects/testing-junit-platform/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestExecutionListener.java

```diff
@@ -134,17 +136,24 @@ private boolean isFailedContainer(TestIdentifier testIdentifier, TestExecutionRe
 
     private TestDescriptorInternal getDescriptor(final TestIdentifier test) {
         if (isMethod(test)) {
-            return new DefaultTestDescriptor(idGenerator.generateId(), className(test), test.getDisplayName());
+            return createDescriptor(test, test.getLegacyReportingName(), test.getDisplayName());
         } else if (isVintageDynamicLeafTest(test)) {
             UniqueId uniqueId = UniqueId.parse(test.getUniqueId());
             return new DefaultTestDescriptor(idGenerator.generateId(), vintageDynamicClassName(uniqueId), vintageDynamicMethodName(uniqueId));
         } else if (isClass(test) || isVintageDynamicTestClass(test)) {
-            return new DefaultTestDescriptor(idGenerator.generateId(), className(test), "classMethod");
+            return createDescriptor(test, "classMethod", "classMethod");
         } else {
-            return new DefaultTestDescriptor(idGenerator.generateId(), className(test), test.getDisplayName());
+            return createDescriptor(test, test.getDisplayName(), test.getDisplayName());
         }
     }
 
+    private TestDescriptorInternal createDescriptor(TestIdentifier test, String name, String displayName) {
+        Optional<TestIdentifier> classIdentifier = findClassSource(test);
```

Comment:
⭕️ Why use optional here instead of just using null?

Reply:
Well, I think that's what `Optional` is designed for - explicitly tells the reader it's possibly be empty - we can sure use `null`, but `Optional` can convey more information IMHO given that we're already using Java 8 in this module.

Reply:
Optional should only be used if null could have a different meaning or is forbidden. Otherwise it just adds more noise to the code.

Reply:
Well I think this says "`findClassSource()` might not be able to find the identifier which contains a class, so you should distinguish these two situations". Basically, I think any scenarios with

```
Object result = getResult()

if(result != null){
     doSomething();
}
```

can and should be replaced with 

```
Optional<Object> result = getResult()

if(result.isPresent()){
     doSomething();
}
```

And I think both JDK8+ and JUnit5 are already practicing this - I can hardly find any new API returning `null` to distinguish existence and unexistence.

WDYT?

Reply:
The identifier of a class is never null, so we don't need Optional to distinguish that. Also, this is not an API, it's a low level private method directly next to where it is used.

Reply:
This conversation sounds like [bikeshedding](https://en.wiktionary.org/wiki/bikeshedding).

I think that what @oehme want's could be achieved by simply rewriting `className` to be:
```java
    private String className(Optional<TestIdentifier> testClassIdentifier) {
        return testClassIdentifier
                .flatMap(TestIdentifier::getSource)
                .map(TestSource::getClassName)
                .orElse("UnknownClass");
    }
```

Reply:
Optional has no place in the private internals of a class. It just adds unnecessary overhead without making it easier to understand. Optional is for public APIs, as Brian Goetz explained in [his answer](https://stackoverflow.com/questions/26327957/should-java-8-getters-return-optional-type/26328555#26328555) on StackOverflow.

Please remove the usage of optional here and let's move on.

Reply:
I'm going to drop this after this, because it's not my code base, and I don't have a hugely vested interest. But I just want to leave you with some of my thoughts.

- I did a quick search through your issues and came up with 31 [Gradle Open & Closed Bugs that have have "NullPointerException"](https://github.com/gradle/gradle/issues?utf8=%E2%9C%93&q=is%3Aissue+NullPointerException+label%3Aa%3Abug)
- There's a reason that Kotlin treats nullable types differently, like a wholly different type. It's because, according to their inventor, they are [The Billion Dollar Mistake](http://en.wikipedia.org/wiki/Tony_Hoare#Apologies_and_retractions). The Kotlin compiler forces you to treat nullable types differently. The compiler doesn't just enforce this when you are exposing a public API (like you are suggesting).
- As a reviewer of code on Github, I have no way of telling if a variable is nullable or not. I don't have the IDE support around me to help me see that "hey, you forgot to check for null here". Having the explicit type gives me the opportunity to clearly see that the lack of presence of a value has been adequately checked.

Personally, I spend most of my day working in Kotlin (but in a mixed Java/Kotlin code base), and the one thing I will say that the downside to `Optional` is that it clashes with the paradigm of Kotlin's nullable types.

Reply:
I think we're violently agreeing here. I'm all for better null handling. I really like how Kotlin solved it.

But Optional was never intended to play that role (please read Brian Goetz comment on this, he's the architect behind it). It was only meant to solve a very specific purpose: As a return type for API methods where using null would almost certainly result in an error. For example, if we could still change it, `Map.get` should have returned Optional, because returning null can mean "there was no entry with this key" and "there was an entry and its value was null".

You can also see that Optional is not the solution to the general nullability problem given some of its other characteristics:

- an Optional variable itself can be null
- an Optional can be present and contain a null

And as you pointed out correctly, using Optional even clashes with better approaches like that of Kotlin, another reason to use it carefully.

Reply:
Good to know, now I realize my mistake and misusing `Optional` before. It's removed now.